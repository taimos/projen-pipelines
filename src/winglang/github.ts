import { Component } from 'projen';
import { GitHubProject, GithubWorkflow } from 'projen/lib/github';
import { JobPermission, JobStep } from 'projen/lib/github/workflows-model';

export interface GithubWinglangPipelineOptions {
  readonly stages: string[];
  readonly target: 'tf-aws';
  readonly wingEntryFile: string;
  readonly iamRoleArnPerStage: { [stage: string]: string };
  readonly awsRegionPerStage: { [stage: string]: string };
}

export class GithubWinglangPipeline extends Component {
  readonly worklfow: GithubWorkflow;
  private readonly currentStages: string[] = [];
  constructor(scope: GitHubProject, protected options: GithubWinglangPipelineOptions) {
    super(scope);

    this.worklfow = scope.github!.addWorkflow('deploy');

    this.worklfow.on({
      push: {
        branches: ['main'],
      },
      workflowDispatch: {},
    });

    this.createCompile();

    for (const stage of options.stages) {
      if (this.options.target === 'tf-aws') {
        this.createTerraformDeploy(stage);
      } else {
        throw new Error(`Unsupported target: ${this.options.target}`);
      }
    }
  }
  createTerraformDeploy(stage: string) {
    const jobSteps: JobStep[] = [
      {
        name: 'Download Wing compile artifact',
        uses: 'actions/download-artifact@v4',
        with: {
          name: 'wing-compile-artifact',
          path: 'target',
        },
      },
      {
        uses: 'aws-actions/configure-aws-credentials@v4',
        with: {
          'aws-region': this.options.awsRegionPerStage[stage],
          'role-to-assume': this.options.iamRoleArnPerStage[stage],
          'role-session-name': `DeploySession-${stage}`,
        },
      },
      {
        name: 'Terrform Init',
        run: 'terraform init',
        workingDirectory: `target/${this.options.wingEntryFile.split('.w')[0]}.tfaws`,
      },
      {
        name: 'Deploy',
        run: 'terraform apply -auto-approve',
        workingDirectory: `target/${this.options.wingEntryFile.split('.w')[0]}.tfaws`,
      },
    ];

    this.worklfow.addJob(`deploy-${stage}`, {
      runsOn: ['ubuntu-latest'],
      needs: ['compile', ...this.currentStages.map(s => `deploy-${s}`)],
      permissions: {
        contents: JobPermission.READ,
        idToken: JobPermission.WRITE,
      },
      steps: jobSteps,
    });
    this.currentStages.push(stage);
  }

  private createCompile() {
    const jobSteps: JobStep[] = [
      {
        name: 'Checkout',
        uses: 'actions/checkout@v4',
      },
      {
        name: 'Install Winglang CLI',
        run: 'npm install -g winglang',
      },
      {
        name: 'Build',
        run: `wing compile - t ${this.options.target} ${this.options.wingEntryFile}`,
      },
      {
        name: 'Upload Wing compile artifact',
        uses: 'actions/upload-artifact@v4',
        with: {
          name: 'wing-compile-artifact',
          path: 'target',
        },
      },
    ];

    this.worklfow.addJob('compile', {
      runsOn: ['ubuntu-latest'],
      permissions: {
        contents: JobPermission.READ,
      },
      steps: jobSteps,
    });

  }
}
