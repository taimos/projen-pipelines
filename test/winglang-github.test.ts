import { GitHubProject } from 'projen/lib/github';
import { synthSnapshot } from 'projen/lib/util/synth';
import { GithubWinglangPipeline } from '../src';

test('Winglang Github snapshot', () => {
  const p = new GitHubProject({
    name: 'testapp',
  });

  new GithubWinglangPipeline(p, {
    stages: ['dev', 'prod'],
    iamRoleArnPerStage: {
      dev: 'devRole',
      prod: 'prodRole',
    },
    awsRegionPerStage: {
      dev: 'eu-central-1',
      prod: 'eu-central-1',
    },
    target: 'tf-aws',
    wingEntryFile: 'main.w',
  });

  const snapshot = synthSnapshot(p);
  expect(snapshot['.github/workflows/deploy.yml']).toMatchSnapshot();
});

