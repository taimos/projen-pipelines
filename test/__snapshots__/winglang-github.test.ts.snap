// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Winglang Github snapshot 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: deploy
on:
  push:
    branches:
      - main
  workflow_dispatch: {}
jobs:
  compile:
    runs-on: ubuntu-latest
    permissions:
      contents: read
    steps:
      - name: Checkout
        uses: actions/checkout@v4
      - name: Install Winglang CLI
        run: npm install -g winglang
      - name: Build
        run: wing compile - t tf-aws main.w
      - name: Upload Wing compile artifact
        uses: actions/upload-artifact@v4
        with:
          name: wing-compile-artifact
          path: target
  deploy-dev:
    needs: compile
    runs-on: ubuntu-latest
    permissions:
      contents: read
      id-token: write
    steps:
      - name: Download Wing compile artifact
        uses: actions/download-artifact@v4
        with:
          name: wing-compile-artifact
          path: target
      - uses: aws-actions/configure-aws-credentials@v4
        with:
          aws-region: eu-central-1
          role-to-assume: devRole
          role-session-name: DeploySession-dev
      - name: Terrform Init
        run: terraform init
        working-directory: target/main.tfaws
      - name: Deploy
        run: terraform apply -auto-approve
        working-directory: target/main.tfaws
  deploy-prod:
    needs:
      - compile
      - deploy-dev
    runs-on: ubuntu-latest
    permissions:
      contents: read
      id-token: write
    steps:
      - name: Download Wing compile artifact
        uses: actions/download-artifact@v4
        with:
          name: wing-compile-artifact
          path: target
      - uses: aws-actions/configure-aws-credentials@v4
        with:
          aws-region: eu-central-1
          role-to-assume: prodRole
          role-session-name: DeploySession-prod
      - name: Terrform Init
        run: terraform init
        working-directory: target/main.tfaws
      - name: Deploy
        run: terraform apply -auto-approve
        working-directory: target/main.tfaws
"
`;
